File: pyfenstein3d/engine/item_grid.py
import math
import numpy as np
from .item import Item
from .vector2d import Vector2d
from .decoration import Decoration
from .wall import Wall
from .door import Door
from .block import Block


class ItemGrid():
    def __init__(self, blocks: []):
        self.__max_x = max(map(lambda i: i.block_x, blocks)) + 1
        self.__max_y = max(map(lambda i: i.block_y, blocks)) + 1
        grid = [None] * self.__max_y * self.__max_x
        self.__items = []
        self.__doors = []
        self.__grid = np.asarray([grid[n:n+self.__max_y]
                                  for n in range(0, len(grid), self.__max_y)])
        for block in blocks:
            if isinstance(block, Item):
                self.__items.append(block)
            if isinstance(block, Door):
                self.__doors.append(block)
            if not block.is_moveable:
                self.__grid[block.block_x, block.block_y] = block

    def get_block(self, block_x, block_y) -> Item:
        block_x = math.floor(block_x)
        block_y = math.floor(block_y)
        if block_x < self.max_x and block_y < self.max_y:
            return self.__grid[block_x, block_y]
        return None

    def remove_block(self, block_x, block_y):
        block_x = math.floor(block_x)
        block_y = math.floor(block_y)
        self.__grid[block_x, block_y] = None

    def get_blocks_by_fov(self, pos: Vector2d, fov_ang: float, fov_delta: float, dist: float, type_block = Block):
        blocks = self.__grid
        fov_ang_vec_min = Vector2d.create_with_ang(fov_ang - (fov_delta * 0.5))
        fov_ang_vec_max = Vector2d.create_with_ang(fov_ang + (fov_delta * 0.5))
        if fov_ang_vec_min.x > 0 and fov_ang_vec_max.x > 0:
            blocks = blocks[math.floor(pos.x):, :]
        elif fov_ang_vec_min.x < 0 and fov_ang_vec_max.x < 0:
            blocks = blocks[:math.floor(pos.x+1), :]
        if fov_ang_vec_min.y > 0 and fov_ang_vec_max.y > 0:
            blocks = blocks[:, math.floor(pos.y):]
        elif fov_ang_vec_min.y < 0 and fov_ang_vec_max.y < 0:
            blocks = blocks[:, :math.floor(pos.y+1)]

        blocks = [block for line in blocks for block in line
                      if isinstance(block, type_block)]

        blocks = [block for block in blocks if block.is_in_fov(pos, fov_ang, fov_delta, dist)]
        
        return blocks
        
    def get_doors_by_fov(self, pos: Vector2d, fov_ang: float, fov_delta: float, dist: float):
        return self.get_blocks_by_fov(pos, fov_ang, fov_delta, dist, Door)

    def get_items_by_fov(self, pos: Vector2d, fov_ang: float, fov_delta: float, dist: float):
        return self.get_blocks_by_fov(pos, fov_ang, fov_delta, dist, Item)

    @property
    def max_x(self):
        return self.__max_x

    @property
    def max_y(self):
        return self.__max_y

File: engine/item_grid.go
package engine

import "math"

// ItemGrid is a 2D spatial grid of blocks.
type ItemGrid struct {
	grid    [][]Blocker
	items   []Blocker
	doors   []*Door
	enemies []*Enemy
	maxX    int
	maxY    int
}

// NewItemGrid creates a new ItemGrid from a list of blocks.
func NewItemGrid(blocks []Blocker) *ItemGrid {
	maxX, maxY := 0, 0
	for _, b := range blocks {
		if b.BlockX() > maxX {
			maxX = b.BlockX()
		}
		if b.BlockY() > maxY {
			maxY = b.BlockY()
		}
	}
	maxX++
	maxY++

	grid := make([][]Blocker, maxX)
	for i := range grid {
		grid[i] = make([]Blocker, maxY)
	}

	ig := &ItemGrid{grid: grid, maxX: maxX, maxY: maxY}

	for _, b := range blocks {
		if _, ok := b.(ItemLike); ok {
			ig.items = append(ig.items, b)
		}
		if door, ok := b.(*Door); ok {
			ig.doors = append(ig.doors, door)
		}
		if enemy, ok := b.(*Enemy); ok {
			ig.enemies = append(ig.enemies, enemy)
		}
		if !b.IsMoveable() {
			ig.grid[b.BlockX()][b.BlockY()] = b
		}
	}
	return ig
}

func (ig *ItemGrid) MaxX() int { return ig.maxX }
func (ig *ItemGrid) MaxY() int { return ig.maxY }

// GetBlock returns the block at the given grid coordinates, or nil if out of bounds.
func (ig *ItemGrid) GetBlock(blockX, blockY int) Blocker {
	bx := int(math.Floor(float64(blockX)))
	by := int(math.Floor(float64(blockY)))
	if bx < ig.maxX && by < ig.maxY && bx >= 0 && by >= 0 {
		return ig.grid[bx][by]
	}
	return nil
}

// RemoveBlock removes the block at the given grid coordinates.
func (ig *ItemGrid) RemoveBlock(x, y float64) {
	bx := int(math.Floor(x))
	by := int(math.Floor(y))
	if bx < ig.maxX && by < ig.maxY && bx >= 0 && by >= 0 {
		ig.grid[bx][by] = nil
	}
}

// GetDoorsByFov returns doors within the field of view.
func (ig *ItemGrid) GetDoorsByFov(pos *Vector2d, fovAng, fovDelta, dist float64) []*Door {
	var result []*Door
	for _, door := range ig.doors {
		if door.IsInFov(pos, fovAng, fovDelta, dist) {
			result = append(result, door)
		}
	}
	return result
}

// GetItemsByFov returns items and enemies within the field of view.
func (ig *ItemGrid) GetItemsByFov(pos *Vector2d, fovAng, fovDelta, dist float64) []Blocker {
	result := ig.getItemsFromGridByFov(pos, fovAng, fovDelta, dist)
	result = append(result, ig.getEnemiesByFov(pos, fovAng, fovDelta, dist)...)
	return result
}

func (ig *ItemGrid) getItemsFromGridByFov(pos *Vector2d, fovAng, fovDelta, dist float64) []Blocker {
	fovAngVecMin := CreateWithAng(fovAng - fovDelta*0.5)
	fovAngVecMax := CreateWithAng(fovAng + fovDelta*0.5)

	startX, endX := 0, ig.maxX
	startY, endY := 0, ig.maxY

	if fovAngVecMin.X() > 0 && fovAngVecMax.X() > 0 {
		startX = int(math.Floor(pos.X()))
	} else if fovAngVecMin.X() < 0 && fovAngVecMax.X() < 0 {
		endX = int(math.Floor(pos.X() + 1))
	}
	if fovAngVecMin.Y() > 0 && fovAngVecMax.Y() > 0 {
		startY = int(math.Floor(pos.Y()))
	} else if fovAngVecMin.Y() < 0 && fovAngVecMax.Y() < 0 {
		endY = int(math.Floor(pos.Y() + 1))
	}

	if startX < 0 {
		startX = 0
	}
	if endX > ig.maxX {
		endX = ig.maxX
	}
	if startY < 0 {
		startY = 0
	}
	if endY > ig.maxY {
		endY = ig.maxY
	}

	var result []Blocker
	for x := startX; x < endX; x++ {
		for y := startY; y < endY; y++ {
			b := ig.grid[x][y]
			if b == nil {
				continue
			}
			if _, ok := b.(ItemLike); !ok {
				continue
			}
			if b.IsInFov(pos, fovAng, fovDelta, dist) {
				result = append(result, b)
			}
		}
	}
	return result
}

func (ig *ItemGrid) getEnemiesByFov(pos *Vector2d, fovAng, fovDelta, dist float64) []Blocker {
	var result []Blocker
	for _, e := range ig.enemies {
		if e.IsInFov(pos, fovAng, fovDelta, dist) {
			result = append(result, e)
		}
	}
	return result
}
